test/performance/scale_test.go

```diff
@@ -0,0 +1,153 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// scale_test.go brings up a number of services tracking the time to various waypoints.
+
+package performance
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/knative/pkg/test/logging"
+	"github.com/knative/test-infra/shared/testgrid"
+
+	"github.com/knative/serving/test/e2e"
+)
+
+type metrics struct {
+	min           time.Duration
+	max           time.Duration
+	totalDuration time.Duration
+	num           int64
+}
+
+func (m *metrics) Add(start time.Time) {
+	// Compute the duration since the provided start time.
+	d := time.Since(start)
+
+	if d < m.min || m.min == 0 {
+		m.min = d
+	}
+	if d > m.max {
+		m.max = d
+	}
+	m.totalDuration += d
+	m.num++
+}
+
+func (m metrics) Min() float32 {
+	return float32(m.min.Seconds())
+}
+
+func (m metrics) Max() float32 {
+	return float32(m.max.Seconds())
+}
+
+func (m metrics) Avg() float32 {
+	return float32(m.totalDuration.Seconds()) / float32(m.num)
+}
+
+type latencies struct {
+	// Guards access to latencies.
+	m       sync.RWMutex
+	metrics map[string]metrics
+}
+
+var _ e2e.Latencies = (*latencies)(nil)
+
+func (l *latencies) Add(name string, start time.Time) {
+	l.m.Lock()
+	defer l.m.Unlock()
+
+	m := l.metrics[name]
+	m.Add(start)
+	l.metrics[name] = m
+}
+
+func (l *latencies) Min(name string) float32 {
+	l.m.RLock()
+	defer l.m.RUnlock()
+	return l.metrics[name].Min()
+}
+
+func (l *latencies) Max(name string) float32 {
+	l.m.RLock()
+	defer l.m.RUnlock()
+	return l.metrics[name].Max()
+}
+
+func (l *latencies) Avg(name string) float32 {
+	l.m.RLock()
+	defer l.m.RUnlock()
+	return l.metrics[name].Avg()
+}
+
+func (l *latencies) Results(t *testing.T) []testgrid.TestCase {
+	l.m.RLock()
+	defer l.m.RUnlock()
+
+	order := make([]string, 0, len(l.metrics))
+	for k := range l.metrics {
+		order = append(order, k)
+	}
+	sort.Strings(order)
+
+	// Add latency metrics
+	tc := make([]testgrid.TestCase, 0, 3*len(order))
+	for _, key := range order {
+		tc = append(tc,
+			CreatePerfTestCase(l.Min(key), fmt.Sprintf("%s.min", key), t.Name()),
+			CreatePerfTestCase(l.Max(key), fmt.Sprintf("%s.max", key), t.Name()),
+			CreatePerfTestCase(l.Avg(key), fmt.Sprintf("%s.avg", key), t.Name()))
+	}
+	return tc
+}
+
+func TestScaleToN(t *testing.T) {
+	// Run each of these variations.
+	tests := []int{10, 50, 100}
+
+	// Accumulate the results from each row in our table (recorded below).
+	var results []testgrid.TestCase
+
+	for _, size := range tests {
+		t.Run(fmt.Sprintf("scale-%d", size), func(t *testing.T) {
+			// Add test case specific name to its own logger
+			logger := logging.GetContextLogger(t.Name())
+
+			// Record the observed latencies.
+			l := &latencies{
+				metrics: make(map[string]metrics),
+			}
+			defer func() {
+				results = append(results, l.Results(t)...)
+			}()
+
+			e2e.ScaleToWithin(t, logger, size, 20*time.Minute, l)
```

Comment:
we keep time fixed here?

Reply:
Checking completion within a reasonable timeframe isn't really the purpose of this test, so I decided against doing what I did in the other scale test.